### ISSUE TITLE
PP project group attribute 

### ISSUE THREAD

**Issue by sacrevert** [BiologicalRecordsCentre/BSBI-Card-and-PlantPortal-DEPRECATED-#63]
This is simply a way of associating plots with a specific project (i.e. there might be several plot groups (and many samples) that are ultimately associated with one project).

It would be a free text field that the user specifies. There would also need to be a check against existing project names when new data are uploaded that use this field.

Nested in #46

**Comment by andrewvanbreda** [comment]
Plot groups have already been setup as terms, therefore if we want to store any information against a plot group it should probably be using the new term attributes.
Now the real question would be how the user enters this information, the most obvious thing would be as an extra column in the input file
My main concern here is from what I can see this adds quite a lot of extra complexity. We would need further data consistency checks, and also no Indicia importer currently understands term attributes. I can see this being an unreasonably large amount of programming for the gain.
So, I am wondering if we scrap the idea of doing this via the import and have the user enter a project for the plot group on a plot group edit screen after the import.
However there is a big question here, how many plot groups would a person typically be importing?
Perhaps we could have a grid actually with checkboxes, the user could select all the plot groups they want a particular project name to apply to.

We know there might be lots of rows, but I am unsure of the number of groups in one import?, the solution might depend on the number of plot groups you are expecting in one go.
Looking at your import file I think there is even only one

**Comment by sacrevert** [comment]
Just to clarify, there is no need for this to be at the plot group level. It would make more sense at the plot or sample level (or even occurrence level). Although, for your edit screen idea, I agree that it would be simpler at the plot group level (because this is higher in the hierarchy, so fewer things to select).

However, is it a requirement for a user to specify a plot group? What if a user doesn't specify a plot group (because all of their plots/samples are within one project and they don't want to specific groupings of plots within that)? Are we currently enforcing plot groups, even if everything is in the same group? @andrewvanbreda

**Comment by andrewvanbreda** [comment]
Yes OK, you might need to decide at what level to store it to match your requirements.
In my opinion  it might be best at sample level. As a plot could be used for several projects, but would there ever be a case of occurrences within one sample being for different projects?
I wouldn't think so, but I am not doing the work out in the field so I am only advising from my computer desk.
If it was just held at the sample level the programming wouldn't be too bad. It would just be a sample attribute on the import, I would probably have to add to the validation rules that check the import, but I wouldn't imagine this would be anything like the complexity of trying to do it with a term attribute during import.

There is currently no enforcement of plot groups, the plot group column can be left out, or it even can be specified with a mixture of plot groups and rows without plot groups if you so wish. The importer is flexible in that regard.
The only thing to not is row a row to be in a plot group it must have its plot group specified.
If everything it to appear in a plot group, that group must currently appear on every row (even if it is the same). There is no way currently to just say I want this group for all rows (like you might for spatial reference system)

**Comment by sacrevert** [comment]
Ok, your comments on plot groups not being essential (as I thought) lead me to think that the sample level would be best. You are right that samples for a single plot would not be in different projects (logically it could happen, but I believe we can enforce this restriction to stop things from getting too complicated).

I actually think plot level would be best. Then if a person was uploading additional samples for an already loaded plot, they would not have to specify the project. So the only additional check needed would be that new samples loaded against existing plot did not try to provide contradictory project information (presumably providing the same project information again, although unnecessary, would not need to throw an error).

For time and coding efficiency, I would be happy to have this as a check-box type screen, where users manually assign plots to projects. Any addition to the upload tool would be a future addition. Note that on a check-box type screen, it would be ideal if there was the option to 'check all' on a display grid, rather than doing all manually (and the display grid should, say, be able to display 100 rows, so there are not too many pages to tab through when assigning plots to projects).

It would also be desirable if users could be assigned to projects (a quick way to give users permissions to add samples to large sets of plots?/analyse samples/etc.) Again, this might be a future extension to the project.

Finally, I suppose it would be good to encourage users to assign plots to a project (a warning message?); and to assign not just a name, but also a free text description field, which could include links to websites, or academic references.

**Comment by andrewvanbreda** [comment]
OK, have read what you have written I can see that it is critical thing here is going to be finding a way of implementing this in a elegant way that results in code that is maintainable.
The idea that projects need to be associated with users concerns me, as users are already assigned to Plot Groups. Could get confusing.
However, perhaps I am missing a trick here.
Would it be possible just to have a project as a type of plot group (at least in code it could be the same even if we call them different things) 
With your suggestion each plot is assigned a project.
However each plot can be associated with 1 or plot groups.
So programmatically there is a bell ringing in my head telling me we should use (roughly) the same code for both of these things.
Plot Groups already have support to be associated with users, so if we make projects like plot groups they would automatically support this too.

**Comment by sacrevert** [comment]
In theory I have no objection to this suggestion. Presumably this would also mean that one plot can be associated with one or more projects? Logically there is no problem with this, although I might decide that we do not promote this (i.e. we might strongly imply in user guidance that plots should only be associated with one project).

Also, although I implied above that it would be good for every plot to be associated with a project, there is not real need for this to be so. We can certainly allow plots to be without a project (presumably this makes the coding easier, as it is in common with the fact that plot groups are not mandatory either). So, the only difference between plot groups and projects is that projects can contain multiple plot groups, but not vice versa (although even this does not need to be rigorously enforced or enforceable).

**Comment by andrewvanbreda** [comment]
Yes, the rules for Plot Groups don't have to be identical, I am just trying to think of an elegant basis from the coding point of view which will mean the Plant Portal project is still understandable and maintainable.
As I say they don't have to be identical, so if you did want to limit to just one project per plot we could certainly do that.
**OLP: I think this might actually simplify things, as it enforces a certain view of projects as the originating piece of research that set-up the plot. (Otherwise people might want to use projects for other purposes, e.g. bringing together different plots for a new piece of research. This might be desirable in the future, but perhaps this could be done in a different way, e.g. using a 'basket' or temporary space for analysis -- prob no need to worry about this now anyway).**

Again, plots wouldn't have to have a project if you didn't want that, plot groups already support this functionality->there doesn't actually have to be a plot group.
**OLP: Yes, fine. I'm happy for plots not to have a project (or plot group).**

Just thinking about your final statement here,
_"So, the only difference between plot groups and projects is that projects can contain multiple plot groups, but not vice versa (although even this does not need to be rigorously enforced or enforceable)."_
I find this sentence slightly confusing to think about as the project is going to associated with the plot not the plot group, at least programatically speaking.
Now if we assume a plot can only have one project, then I think you are saying that actually a plot group can only ever be used by one project.
Given that, then we could say projects can have multiple plot groups, and plot groups would have one project.
This is something we could (if we wished) enforce in the user interface as the data structure would not naturally be limited in such a way.
Phew, I think I am correct in what I have just said, difficult one to get head around.
**OLP: Yes, I agree with this. Good to hear that it would be a user interface restriction. I suppose this makes it more flexible if changes were needed in the future.**

**Comment by andrewvanbreda** [comment]
@sacrevert Yes this all seems fine

**Comment by andrewvanbreda** [comment]
@sacrevert given that you have now indicated that you wish to have projects ready for the next build, can you confirm that you are happy with the implementation ideas (in above comments) for projects to go ahead.
So essentially a project would work like a plot group in the data structure but I would add code to the user interface to make sure only one project can be associated with each plot.
Hopefully I can re-use a lot of the plot group code, although I still expect quite a lot of effort, even if the coding ends up being simpler than I think will be, the testing effort with be significant
Given this will be significant I just wanted to double-check with you that I am on the right lines with this, also if I know exactly what I am trying to achieve I can provide a better estimate for David.

**Comment by sacrevert** [comment]
@andrewvanbreda @DavidRoy I am happy with the ideas for implementation arrived at at the end of this thread. (These more or less mirror the 'projectData' table that I had originally conceived of when I revised the specification.) I presume then, that the next steps are:

- Andy to estimate the time requirement for implementing the above for the first build
- David to confirm that he is happy for this to be included now, and not later.

If we go ahead now, then I presume that the simplest solution is for the projects to have only one piece of associated information, which is a free text field (rather than a more structured, table, with additional info such as, e.g., description, project code, temporal coverage, organisation etc.)

Andy has previously suggested that project group associations could also be tacked on subsequently, with users selecting plots through a check box system. I presume that the  'on import' solution being suggested requires less resource, but, the advantage of the check box system is that it could be left until we have more funding, then, once implemented, I would need to go through and assign all uploaded (historic) plots to specific projects.

@DavidRoy to comment.

**Comment by andrewvanbreda** [comment]
This has now been discussed over email.

The original suggestion was to make Projects like Plot Groups. This was to get around complications of using essentially what is a sample/occurrence importer with location attributes (as plot groups already overcome these difficulties).
The system needs to be clever enough to know when a new group needs adding to a plot and when to use an existing one.
Also it is worth noting that plot groups can be associated with users, this is a discardable requirement for projects, although not essential for now.

Coding projects to work like plot groups has the disadvantage of requiring quite a lot of coding resource even though we are not starting from scratch.

To overcome this, I suggested using a single location attribute forgetting the complications involved.
It was pointed out by Oli that we had already discussed complications involved with this and it would not be straightforward.

AVB suggested using a sample attribute but again there are issues involved with this. Although it is simple to implement, it has the disadvantage is that it would be difficult to move to a plot level (location) attribute later on.

Given this discussion, it was decided to pursue the following approach
- Do not change the importer at all for now
- Then later on, as required, add a screen to the website which will allow users to apply plots to a particular project after import is complete.
This has the advantage that we are using location attribute, we don’t need to do any coding now, and is highly scalable in that we can extend the import later on to allow projects to work very much like plot groups, including with associated users.

So, for now there is nothing to do on this until the importer is running.
